File: app/models/descriptor.rb
=begin

Copyright (c) 2022, all rights reserved.

All the information provided by this platform is protected by international laws related  to 
industrial property, intellectual property, copyright and relative international laws. 
All intellectual or industrial property rights of the code, texts, trade mark, design, 
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// · 
=end

class Descriptor < ApplicationLesliRecord
    belongs_to :account,  foreign_key: "accounts_id",    class_name: "Account"
    has_many :privileges, foreign_key: "descriptors_id"
    has_many :activities, foreign_key: "descriptors_id"

    # this scope is needed to allow to join with deleted descriptors
    # join with deleted descriptors is needed to know which privileges we have to remove from the
    # role_privileges table when a descriptor is removed from role_describers
    has_many :role_descriptors_all, -> { with_deleted }, foreign_key: "descriptors_id", class_name: "Role::Descriptor"

    validates :name, presence: true

    after_create :initialize_descriptor_privileges

    def initialize_descriptor_privileges

        # assign ["show", "update"] actions from profile descriptor to the role
        DescriptorService.add_profile_privileges(self) if self.name == "profile"

        DescriptorService.add_owner_privileges(self) if ["owner", "sysadmin"].include?(self.name)

    end

    def self.list(current_user, query)
        current_user.account.descriptors
        .select(:id, :name)
        .order(:name)
    end 

    def self.index(current_user, query)

        # Get search string from query params
        search_string = LC::Sql.sanitize_for_search(query[:search])

        descriptors = current_user.account.descriptors.select(
            :id,
            :name,
            :description,
            "coalesce(actions.total, 0) as privileges_count",
            LC::Date2.new.date_time.db_timestamps("descriptors")
        )

        # Count the amount of privileges assigned to every descriptor
        descriptors = descriptors.joins("
            left join  (
                select
                    count(1) as total,
                    descriptors_id
                from descriptor_privileges
                --where apga.status = TRUE
                group by descriptors_id
            ) as actions
                on actions.descriptors_id = descriptors.id
        ")
        
        # skip native descriptors
        #descriptors = descriptors.where.not("descriptors.name in (?)", ["owner", "sysadmin", "profile"])

        # Filter results by search string
        unless search_string.blank?
            descriptors = descriptors.where("(LOWER(descriptors.name) SIMILAR TO ?)", search_string)
        end

        descriptors
        .page(query[:pagination][:page])
        .per(query[:pagination][:perPage])
        .order("#{query[:order][:by]} #{query[:order][:dir]} NULLS LAST")

    end 

    def show(current_user, query)
        { 
            :id => self.id,
            :name => self.name,
            :description => self.description,
            :descriptors_id => self.descriptors_id,
            :privileges => self.privileges.select(:id, :controller, :action, :created_at)
        }
    end
    
end

File: app/vue3/administration/stores/descriptor.js
/*
Copyright (c) 2022, all rights reserved.

All the information provided by this platform is protected by international laws related  to 
industrial property, intellectual property, copyright and relative international laws. 
All intellectual or industrial property rights of the code, texts, trade mark, design, 
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// · 
*/


// · 
import { defineStore } from "pinia"


// · 
export const useDescriptor = defineStore("administration.descriptor", {
    state: () => {
        return {
            list: [],
            records: [],
            pagination: {},
            descriptor: {},
            privileges: {},
            loading: false,
            pagination: {
                page: 1
            },
            index: { 
                pagination: {},
                records: []
            },
            descriptors_options: []
        }
    },
    actions: {

        fetch(url=this.url.admin("descriptors")) {
            this.loading = true
            this.http.get(url).then(result => {
                this.index = result
                this.records = result.records
                this.loading = false
            })
        },

        fetchDescriptor(id) {
            this.http.get(this.url.admin("descriptors/:id", id)).then(result => {
                this.descriptor = result
            })
        },

        fetchDescriptorList() {
            this.http.get(this.url.admin("descriptors/list")).then(result => {
                this.list = result.map(descriptor => {
                    descriptor.active = false
                    return descriptor
                })
            })
        },
        /**
        * @description This action is used to reset descriptor object
        */
        resetDescriptor(){
            this.descriptor = {}
        },
        /**
        * @description This action is used to create a descriptor
        */
        createDescriptor(){
            this.loading = true
            this.http.post(this.url.admin("descriptors"), { descriptor: this.descriptor }).then(result => {
                this.descriptor.id = result.id
                this.msg.success(I18n.t("core.users.messages_success_operation"))
                this.loading = false
            }).catch(error => {
                this.msg.danger(I18n.t("core.shared.messages_danger_internal_error"))
            })

        },
        /**
        * @description This action is used to update a descriptor
        */
        updateDescriptor(){
            this.loading = true
            this.http.put(this.url.admin("descriptors/:id", this.descriptor.id), { descriptor: this.descriptor }).then(result => {
                this.msg.success(I18n.t("core.users.messages_success_operation"))
                this.loading = false
            }).catch(error => {
                this.msg.danger(I18n.t("core.shared.messages_danger_internal_error"))
            })

        },
        /**
        * @description This action is used to sort the results
        */
        sortIndex(column, direction) {
            this.fetch(this.url.admin("descriptors").order(column, direction))
        },
        /**
         * @description This action is used to search
         * @param {string} search_string
         */
        search(search_string) {
            this.fetch(this.url.admin("descriptors").search(search_string))
        },
        /**
         * @description This action is used to paginate index
         * @param {string} page actual page 
         */
        paginateIndex(page) {
            this.pagination.page = page
            this.fetch(this.url.admin("descriptors").paginate(this.pagination.page))
        },
        /**
        * @description This action is used to get descriptors as options
        */
        getDescriptorsOptions(){
            this.loading = true
            this.descriptors_options = []
            this.http.get(this.url.admin("descriptors/list")).then(result => {
                result.forEach((descriptor)=>{
                    this.descriptors_options.push({
                        label: descriptor.name,
                        value: descriptor.id
                    })
                })
                
                this.loading = false
            })
        },

        fetchDescriptorPrivileges() {
            this.http.get(this.url.admin("descriptors/:id/privileges", this.descriptor.id)).then(result => {
                result.forEach(controllerAction => {
                    if (!this.privileges[controllerAction.action]) {
                        this.privileges[controllerAction.action] = []
                    }
                    this.privileges[controllerAction.action].push(controllerAction)
                })
            })
        },


        // Add privilege to descriptor
        postPrivilege(action) {
            this.http.post(this.url.admin("descriptors/:id/privileges", this.descriptor.id), {
                descriptor_privilege: {
                    controller_id: action.controller_id,
                    action_id: action.action_id
                }
            }).then(result => {
                console.log(result)
            }).catch(error => {
                console.log(error)
            })
        },


        // Add privilege to descriptor
        deletePrivilege(action) {
            this.http.delete(this.url.admin("descriptors/:id/privileges", this.descriptor.id), {
                descriptor_privilege: {
                    controller_id: action.controller_id,
                    action_id: action.action_id
                }
            }).then(result => {
                console.log(result)
            }).catch(error => {
                console.log(error)
            })
        }

    }
})
